File: src/follow_target_python/follow_target_python/orgwoa_lidar.py
#!/usr/bin/env python3

import math
import numpy as np
import time 
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan


class WOA_Follow(Node):

    def __init__(self):
        super().__init__('woa_follow')

        # ---------------- Parameters ----------------
        self.d_ref = 0.7
        self.stop_threshold = 0.05

        self.v_max = 0.6
        self.w_max = 3.0

        self.alpha = 2.0
        self.beta = 1.5

        self.n_whales = 25
        self.max_iter = 10

        self.dt = 0.6
        self.b = 1.0

        # 🔥 ROBOT SIZE
        self.robot_radius = 0.22
        self.safety_margin = 0.05

        # ---------- Detection / tracking ----------
        self.track_gate = 0.50
        self.max_lost_cycles = 6
        self.lost_cycles = 0
        self.prev_target = None

        # Smoothing
        self.prev_v = 0.0
        self.prev_w = 0.0
        self.smooth_gain = 0.2

        # ---------------- ROS ----------------
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel', 10)

        self.scan = None
        self.create_subscription(LaserScan, '/scan', self.scan_callback, 10)

        self.timer = self.create_timer(0.1, self.control_loop)

        self.get_logger().info("WOA follower with HARD safety + LiDAR tracking")

    # --------------------------------------------------
    def scan_callback(self, msg):
        self.scan = msg

    # --------------------------------------------------
    # ----------- LIDAR PROCESSING ---------------------
    # --------------------------------------------------

    def scan_to_points(self):
        points = []
        for i, r in enumerate(self.scan.ranges):

            if math.isinf(r) or math.isnan(r):
                continue

            angle = self.scan.angle_min + i * self.scan.angle_increment

            if abs(angle) > 1.2:
                continue

            x = r * math.cos(angle)
            y = r * math.sin(angle)

            points.append((x, y))

        return points

    def cluster_points(self, points, threshold=0.15):
        clusters = []
        used = [False]*len(points)

        for i in range(len(points)):
            if used[i]:
                continue

            cluster = [points[i]]
            used[i] = True
            queue = [points[i]]

            while queue:
                px, py = queue.pop()

                for j in range(len(points)):
                    if used[j]:
                        continue

                    qx, qy = points[j]
                    if math.hypot(px - qx, py - qy) < threshold:
                        used[j] = True
                        cluster.append((qx, qy))
                        queue.append((qx, qy))

            clusters.append(cluster)

        return clusters

    def compute_cluster_features(self, cluster):
        xs = [p[0] for p in cluster]
        ys = [p[1] for p in cluster]

        cx = np.mean(xs)
        cy = np.mean(ys)

        width = max(ys) - min(ys)
        length = max(xs) - min(xs)

        return cx, cy, width, length

    def select_target(self, clusters):

        candidates = []

        for cluster in clusters:
            if len(cluster) < 5:
                continue

            cx, cy, w, l = self.compute_cluster_features(cluster)

            size = math.hypot(w, l)

            if size < 0.08:
                continue

            if size > 0.6:
                continue

            if not (0.08 < w < 0.35 or 0.08 < l < 0.35):
                continue

            candidates.append((cx, cy))

        if not candidates:
            if self.prev_target is not None and self.lost_cycles < self.max_lost_cycles:
                self.lost_cycles += 1
                return self.prev_target
            return None

        if self.prev_target is None:
            self.lost_cycles = 0
            return min(candidates, key=lambda p: math.hypot(p[0], p[1]))

        px, py = self.prev_target

        gated_candidates = [
            p for p in candidates
            if math.hypot(p[0] - px, p[1] - py) < self.track_gate
        ]

        if gated_candidates:
            self.lost_cycles = 0
            return min(gated_candidates, key=lambda p: math.hypot(p[0] - px, p[1] - py))

        if self.lost_cycles < self.max_lost_cycles:
            self.lost_cycles += 1
            return self.prev_target

        self.lost_cycles = 0
        return min(candidates, key=lambda p: math.hypot(p[0], p[1]))

    # --------------------------------------------------
    # 🔥 HARD COLLISION CHECK
    # --------------------------------------------------

    def is_safe(self, v, w):

        if self.scan is None:
            return True

        theta = w * self.dt
        new_x = v * math.cos(theta)
        new_y = v * math.sin(theta)

        for i, r in enumerate(self.scan.ranges):

            if math.isinf(r) or math.isnan(r):
                continue

            angle = self.scan.angle_min + i * self.scan.angle_increment

            if abs(angle) > math.pi / 3:
                continue

            obs_x = r * math.cos(angle)
            obs_y = r * math.sin(angle)

            dist = math.hypot(new_x - obs_x, new_y - obs_y)

            if dist < (self.robot_radius + self.safety_margin):
                return False

        return True

    # --------------------------------------------------
    def predict_relative_state(self, dx, dy, v, w):

        x_rel = dx - v * self.dt
        y_rel = dy

        th = -w * self.dt
        c = math.cos(th)
        s = math.sin(th)

        dx_next = c * x_rel - s * y_rel
        dy_next = s * x_rel + c * y_rel

        return dx_next, dy_next

    # --------------------------------------------------
    def fitness(self, candidate, dx, dy):

        v = float(candidate[0])
        w = float(candidate[1])

        if not self.is_safe(v, w):
            return float('inf')

        dx_p, dy_p = self.predict_relative_state(dx, dy, v, w)

        d = math.hypot(dx_p, dy_p)
        theta = math.atan2(dy_p, dx_p)

        e_d = abs(d - self.d_ref)
        e_theta = abs(theta)

        effort = 0.05 * abs(v) + 0.02 * abs(w)

        return self.alpha * e_d + self.beta * e_theta + effort

    # --------------------------------------------------
    def woa_optimize(self, dx, dy):

        lb = np.array([-self.v_max, -self.w_max])
        ub = np.array([ self.v_max,  self.w_max])

        whales = np.random.uniform(lb, ub, size=(self.n_whales, 2))

        fitness_vals = np.array([self.fitness(w, dx, dy) for w in whales])

        if np.all(np.isinf(fitness_vals)):
            return 0.0, 0.8

        best_idx = np.argmin(fitness_vals)
        X_best = whales[best_idx].copy()
        F_best = fitness_vals[best_idx]

        for t in range(self.max_iter):

            a = 2.0 - 2.0 * (t / float(self.max_iter))

            for i in range(self.n_whales):

                r1 = np.random.rand(2)
                r2 = np.random.rand(2)

                A = 2.0 * a * r1 - a
                C = 2.0 * r2

                p = np.random.rand()
                l = np.random.uniform(-1.0, 1.0)

                Xi = whales[i].copy()

                if p < 0.5:
                    if np.linalg.norm(A) < 1.0:
                        D = np.abs(C * X_best - Xi)
                        X_new = X_best - A * D
                    else:
                        rand_idx = np.random.randint(self.n_whales)
                        X_rand = whales[rand_idx].copy()
                        D = np.abs(C * X_rand - Xi)
                        X_new = X_rand - A * D
                else:
                    D_prime = np.abs(X_best - Xi)
                    X_new = D_prime * np.exp(self.b * l) * np.cos(2.0 * math.pi * l) + X_best

                whales[i] = np.clip(X_new, lb, ub)

            fitness_vals = np.array([self.fitness(w, dx, dy) for w in whales])

            if np.all(np.isinf(fitness_vals)):
                return 0.0, 0.8

            current_best_idx = np.argmin(fitness_vals)

            if fitness_vals[current_best_idx] < F_best:
                F_best = fitness_vals[current_best_idx]
                X_best = whales[current_best_idx].copy()

        return float(X_best[0]), float(X_best[1])

    # --------------------------------------------------
    def control_loop(self):
        start_time = time.perf_counter()   


        if self.scan is None:
            return

        points = self.scan_to_points()
        clusters = self.cluster_points(points)
        target = self.select_target(clusters)

        if target is None:
            self.get_logger().warn("Leader not detected")
            return

        self.prev_target = target

        dx, dy = target

        distance = math.hypot(dx, dy)
        heading = math.atan2(dy, dx)

        if abs(distance - self.d_ref) < self.stop_threshold and abs(heading) < 0.08:
            v_cmd = 0.0
            w_cmd = 0.0
        else:
            v_cmd, w_cmd = self.woa_optimize(dx, dy)

        v = (1.0 - self.smooth_gain) * v_cmd + self.smooth_gain * self.prev_v
        w = (1.0 - self.smooth_gain) * w_cmd + self.smooth_gain * self.prev_w

        self.prev_v = v
        self.prev_w = w

        cmd = Twist()
        cmd.linear.x = float(np.clip(v, -self.v_max, self.v_max))
        cmd.angular.z = float(np.clip(w, -self.w_max, self.w_max))
        self.cmd_pub.publish(cmd)
        elapsed_ms = (time.perf_counter() - start_time) * 1000.0

        self.get_logger().info(
            f"FULL LOOP = {elapsed_ms:.2f} ms | "
            f"dist={distance:.2f} v={cmd.linear.x:.2f} w={cmd.angular.z:.2f}"
        )
        self.get_logger().info(
            f"dist={distance:.2f} v={cmd.linear.x:.2f} w={cmd.angular.z:.2f}"
        )
        # Python WOA + LiDAR timing:

        # - Min: ~2 ms (steady state / no movement)
        # - Max: ~335 ms (rare spike)
        # - Avg: ~58–60 ms
        
        # → Usually real-time safe for 10 Hz (100 ms)
        # → BUT NOT deterministic (spikes exist)
        # → Occasional deadline miss (very important)

        # → CPU usage fluctuates depending on WOA convergence


def main(args=None):
    rclpy.init(args=args)
    node = WOA_Follow()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/follow_target_cpp/src/org_woa_lidar.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>
#include <random>
#include <numeric>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

using namespace std::chrono_literals;

struct Point {
    double x, y;
};

class WOAFollow : public rclcpp::Node {
public:
    WOAFollow() : Node("woa_follow"), gen_(rd_()) {
        // ---------------- Parameters (Identical to Python) ----------------
        d_ref_ = 0.7;
        stop_threshold_ = 0.05;
        v_max_ = 0.6;
        w_max_ = 3.0;
        alpha_ = 2.0;
        beta_ = 1.5;
        n_whales_ = 25;
        max_iter_ = 10;
        dt_ = 0.4;
        b_ = 1.0;

        robot_radius_ = 0.22;
        safety_margin_ = 0.05;

        track_gate_ = 0.50;
        max_lost_cycles_ = 6;
        lost_cycles_ = 0;
        prev_target_set_ = false;

        prev_v_ = 0.0;
        prev_w_ = 0.0;
        smooth_gain_ = 0.2;

        cmd_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
        scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10, std::bind(&WOAFollow::scan_callback, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(100ms, std::bind(&WOAFollow::control_loop, this));
        RCLCPP_INFO(this->get_logger(), "WOA follower with HARD safety + LiDAR tracking (Vectorized)");
    }

private:
    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg) { scan_ = msg; }

    // LIDAR PROCESSING
    std::vector<Point> scan_to_points() {
        std::vector<Point> points;
        if (!scan_) return points;
        for (size_t i = 0; i < scan_->ranges.size(); ++i) {
            double r = scan_->ranges[i];
            if (std::isinf(r) || std::isnan(r)) continue;
            double angle = scan_->angle_min + i * scan_->angle_increment;
            if (std::abs(angle) > 1.2) continue;
            points.push_back({r * std::cos(angle), r * std::sin(angle)});
        }
        return points;
    }

    std::vector<std::vector<Point>> cluster_points(const std::vector<Point>& points) {
        std::vector<std::vector<Point>> clusters;
        std::vector<bool> used(points.size(), false);
        for (size_t i = 0; i < points.size(); ++i) {
            if (used[i]) continue;
            std::vector<Point> cluster;
            std::vector<size_t> queue = {i};
            used[i] = true;
            size_t head = 0;
            while (head < queue.size()) {
                Point p = points[queue[head++]];
                cluster.push_back(p);
                for (size_t j = 0; j < points.size(); ++j) {
                    if (!used[j] && std::hypot(p.x - points[j].x, p.y - points[j].y) < 0.15) {
                        used[j] = true; queue.push_back(j);
                    }
                }
            }
            clusters.push_back(cluster);
        }
        return clusters;
    }

    Point select_target(const std::vector<std::vector<Point>>& clusters) {
        std::vector<Point> candidates;
        for (const auto& cluster : clusters) {
            if (cluster.size() < 5) continue;
            double sum_x = 0, sum_y = 0, min_x = 1e6, max_x = -1e6, min_y = 1e6, max_y = -1e6;
            for (const auto& p : cluster) {
                sum_x += p.x; sum_y += p.y;
                min_x = std::min(min_x, p.x); max_x = std::max(max_x, p.x);
                min_y = std::min(min_y, p.y); max_y = std::max(max_y, p.y);
            }
            double w = max_y - min_y, l = max_x - min_x;
            double size = std::hypot(w, l);
            if (size >= 0.08 && size <= 0.6 && ((w > 0.08 && w < 0.35) || (l > 0.08 && l < 0.35))) {
                candidates.push_back({sum_x / cluster.size(), sum_y / cluster.size()});
            }
        }

        if (candidates.empty()) {
            if (prev_target_set_ && lost_cycles_ < max_lost_cycles_) {
                lost_cycles_++; return prev_target_;
            }
            return {0.0, 0.0};
        }

        Point best;
        if (!prev_target_set_) {
            lost_cycles_ = 0;
            best = candidates[0];
            double min_d = std::hypot(best.x, best.y);
            for(auto& c : candidates) {
                double d = std::hypot(c.x, c.y);
                if(d < min_d){ min_d = d; best = c; }
            }
        } else {
            std::vector<Point> gated;
            for(auto& c : candidates) {
                if(std::hypot(c.x - prev_target_.x, c.y - prev_target_.y) < track_gate_) gated.push_back(c);
            }
            if (!gated.empty()) {
                lost_cycles_ = 0;
                best = gated[0];
                double min_d = std::hypot(best.x - prev_target_.x, best.y - prev_target_.y);
                for(auto& g : gated) {
                    double d = std::hypot(g.x - prev_target_.x, g.y - prev_target_.y);
                    if(d < min_d){ min_d = d; best = g; }
                }
            } else if (lost_cycles_ < max_lost_cycles_) {
                lost_cycles_++; return prev_target_;
            } else {
                lost_cycles_ = 0;
                best = candidates[0];
                double min_d = std::hypot(best.x, best.y);
                for(auto& c : candidates) {
                    double d = std::hypot(c.x, c.y);
                    if(d < min_d){ min_d = d; best = c; }
                }
            }
        }
        return best;
    }

    // HARD COLLISION CHECK (Identical to Python is_safe)
    bool is_safe(double v, double w) {
        if (!scan_) return true;
        double theta = w * dt_;
        double new_x = v * std::cos(theta);
        double new_y = v * std::sin(theta);
        for (size_t i = 0; i < scan_->ranges.size(); ++i) {
            double r = scan_->ranges[i];
            if (std::isinf(r) || std::isnan(r)) continue;
            double angle = scan_->angle_min + i * scan_->angle_increment;
            if (std::abs(angle) > M_PI / 3.0) continue;
            double obs_x = r * std::cos(angle);
            double obs_y = r * std::sin(angle);
            if (std::hypot(new_x - obs_x, new_y - obs_y) < (robot_radius_ + safety_margin_)) return false;
        }
        return true;
    }

    double fitness(double v, double w, double dx, double dy) {
        if (!is_safe(v, w)) return 1e12;
        double x_rel = dx - v * dt_;
        double th = -w * dt_;
        double dx_p = std::cos(th) * x_rel - std::sin(th) * dy;
        double dy_p = std::sin(th) * x_rel + std::cos(th) * dy;
        return alpha_ * std::abs(std::hypot(dx_p, dy_p) - d_ref_) + beta_ * std::abs(std::atan2(dy_p, dx_p)) + (0.05 * std::abs(v) + 0.02 * std::abs(w));
    }

    // WOA OPTIMIZATION (100% Python Logic with Vectorized A/C)
    std::pair<double, double> woa_optimize(double dx, double dy) {
        std::uniform_real_distribution<double> dist_v(-v_max_, v_max_);
        std::uniform_real_distribution<double> dist_w(-w_max_, w_max_);
        std::uniform_real_distribution<double> dist_01(0.0, 1.0);
        std::uniform_real_distribution<double> dist_m11(-1.0, 1.0);

        struct Whale { double v, w, fit; };
        std::vector<Whale> whales(n_whales_);
        Whale best_whale = {0, 0, 1e12};

        for (int i = 0; i < n_whales_; ++i) {
            whales[i].v = dist_v(gen_); whales[i].w = dist_w(gen_);
            whales[i].fit = fitness(whales[i].v, whales[i].w, dx, dy);
            if (whales[i].fit < best_whale.fit) best_whale = whales[i];
        }

        if (best_whale.fit >= 1e12) return {0.0, 0.8};

        for (int t = 0; t < max_iter_; ++t) {
            double a = 2.0 - 2.0 * (static_cast<double>(t) / max_iter_);
            for (int i = 0; i < n_whales_; ++i) {
                double p = dist_01(gen_), l = dist_m11(gen_);
                
                // Vectorized A and C (independent randoms for v and w)
                double Av = 2.0 * a * dist_01(gen_) - a;
                double Aw = 2.0 * a * dist_01(gen_) - a;
                double Cv = 2.0 * dist_01(gen_);
                double Cw = 2.0 * dist_01(gen_);

                if (p < 0.5) {
                    if (std::hypot(Av, Aw) < 1.0) { // np.linalg.norm(A) < 1.0
                        whales[i].v = best_whale.v - Av * std::abs(Cv * best_whale.v - whales[i].v);
                        whales[i].w = best_whale.w - Aw * std::abs(Cw * best_whale.w - whales[i].w);
                    } else {
                        int r_idx = std::uniform_int_distribution<int>(0, n_whales_ - 1)(gen_);
                        whales[i].v = whales[r_idx].v - Av * std::abs(Cv * whales[r_idx].v - whales[i].v);
                        whales[i].w = whales[r_idx].w - Aw * std::abs(Cw * whales[r_idx].w - whales[i].w);
                    }
                } else {
                    double spiral = std::exp(b_ * l) * std::cos(2.0 * M_PI * l);
                    whales[i].v = std::abs(best_whale.v - whales[i].v) * spiral + best_whale.v;
                    whales[i].w = std::abs(best_whale.w - whales[i].w) * spiral + best_whale.w;
                }
                whales[i].v = std::clamp(whales[i].v, -v_max_, v_max_);
                whales[i].w = std::clamp(whales[i].w, -w_max_, w_max_);
                whales[i].fit = fitness(whales[i].v, whales[i].w, dx, dy);
            }
            for(auto& w : whales) if(w.fit < best_whale.fit) best_whale = w;
            if (best_whale.fit >= 1e12) return {0.0, 0.8};
        }
        return {best_whale.v, best_whale.w};
    }

    void control_loop() {
        if (!scan_) return;
        auto points = scan_to_points();
        Point target = select_target(cluster_points(points));
        if (target.x == 0.0 && target.y == 0.0 && !prev_target_set_) {
            RCLCPP_WARN(this->get_logger(), "Leader not detected"); return;
        }
        prev_target_ = target; prev_target_set_ = true;
        double dist = std::hypot(target.x, target.y), head = std::atan2(target.y, target.x);
        double v_cmd = 0.0, w_cmd = 0.0;
        if (!(std::abs(dist - d_ref_) < stop_threshold_ && std::abs(head) < 0.08)) {
            auto start = std::chrono::high_resolution_clock::now();   // <-- ADD THIS
            auto res = woa_optimize(target.x, target.y);
            auto end = std::chrono::high_resolution_clock::now();     // <-- ADD THIS
            double solve_time_ms = std::chrono::duration<double, std::milli>(end - start).count();  // <-- ADD THIS
            RCLCPP_INFO(this->get_logger(), "WOA solve time = %.3f ms", solve_time_ms);  // <-- ADD THIS
            v_cmd = res.first; w_cmd = res.second;
        }
        prev_v_ = (1.0 - smooth_gain_) * v_cmd + smooth_gain_ * prev_v_;
        prev_w_ = (1.0 - smooth_gain_) * w_cmd + smooth_gain_ * prev_w_;
        auto cmd = geometry_msgs::msg::Twist();
        cmd.linear.x = std::clamp(prev_v_, -v_max_, v_max_);
        cmd.angular.z = std::clamp(prev_w_, -w_max_, w_max_);
        cmd_pub_->publish(cmd);
        RCLCPP_INFO(this->get_logger(), "dist=%.2f v=%.2f w=%.2f", dist, cmd.linear.x, cmd.angular.z);
    }

    double d_ref_, stop_threshold_, v_max_, w_max_, alpha_, beta_, dt_, b_, smooth_gain_, prev_v_, prev_w_, robot_radius_, safety_margin_, track_gate_;
    int n_whales_, max_iter_, lost_cycles_, max_lost_cycles_;
    bool prev_target_set_; Point prev_target_;
    sensor_msgs::msg::LaserScan::SharedPtr scan_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_pub_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::random_device rd_; std::mt19937 gen_;
    // WOA computation time:
    // - Min: 5.14 ms
    // - Max: 10.89 ms
    // - Avg: 7.35 ms
    //
    // → Real-time safe for 10 Hz control loop (100 ms)
    // → CPU usage ≈ 7–11%
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<WOAFollow>());
    rclcpp::shutdown();
    return 0;
}